File: snippets/ts/src/foundations/services/object.ts
import * as restate from "@restatedev/restate-sdk";

interface Item {
  productId: string;
  quantity: number;
  price: number;
}

// <start_here>
const cartObject = restate.object({
  name: "ShoppingCart",
  handlers: {
    addItem: async (ctx: restate.ObjectContext, item: Item) => {
      const items = (await ctx.get<Item[]>("cart")) ?? [];
      items.push(item);
      ctx.set("cart", items);
      return items;
    },

    getTotal: restate.handlers.object.shared(
      async (ctx: restate.ObjectSharedContext) => {
        const items = (await ctx.get<Item[]>("cart")) ?? [];
        return items.reduce((sum, item) => sum + item.price * item.quantity, 0);
      }
    ),
  },
});
// <end_here>

export default cartObject;

File: snippets/python/src/foundations/services/object.py
import restate
from pydantic import BaseModel
from restate import ObjectContext, ObjectSharedContext
from typing import List, Optional


class Item(BaseModel):
    product_id: str
    quantity: int
    price: float


class Cart(BaseModel):
    items: List[Item] = []


# <start_here>
cart_object = restate.VirtualObject("ShoppingCart")


@cart_object.handler()
async def add_item(ctx: ObjectContext, item: Item) -> Cart:
    cart = await ctx.get("cart", type_hint=Cart) or Cart()
    cart.items.append(item)
    ctx.set("cart", cart)
    return cart


@cart_object.handler(kind="shared")
async def get_total(ctx: ObjectSharedContext) -> float:
    cart = await ctx.get("cart", type_hint=Cart) or Cart()
    return sum(item.price * item.quantity for item in cart.items)


# <end_here>
